Add ability to take screenshots

m1ga opened the pull request:
I've started to add the screenshot (at the moment only pictures, screencast after that) functionality. 

![screenshot_20180701_203217](https://user-images.githubusercontent.com/4334997/42137527-048f4196-7d6e-11e8-8811-9e3d86eacccc.png)

### Current idea:
* try to take as many screenshots as possible: 
* check if it's an ios platform -> try to take iOS screenshot
* check if there are emulators or devices -> try to take Android screenshot

It is working for Android with a connected device. When no device is connected it will show an error.

### Improvements:
* how to handle multiple devices (e.g. use the Android device list)
* Android: remove empty file if an error appeared (e.g. no device connected, but one emulator that is not running)

hansemannn commented on lib/ui/toolbar.js:
```diff
@@ -236,7 +236,8 @@ export default class Toolbar {
 
-						<Button icon="gear" flat="true" disabled={this.state.disableUI || !this.state.codeSigningAvailable} click={this.expandButtonClicked.bind(this)} />
+						<Button title="Compile" icon="gear" flat="true" disabled={this.state.disableUI || !this.state.codeSigningAvailable} click={this.expandButtonClicked.bind(this)} />
+						<Button title="Take screenshot" icon="device-camera" flat="true" disabled={this.state.disableUI} click={this.takeScreenshotButtonClicked.bind(this)}/>
```
I'm not sure taking a screenshot is a prominent enough action to have an own - persistent - button in the UI. What do you think about making it like the clean command / flow?

m1ga replied:
probably not. Removed the button and put it into the menu

hansemannn commented on lib/index.js:
```diff
@@ -523,6 +536,26 @@ export default {
+
+		if (os.platform() === 'darwin') {
+			// ios only
+			// xcrun simctl io booted screenshot
```
Will you implement this part as well / do you have access to a Xcode env?

m1ga replied:
yes, I'll give it a try next week!

m1ga commented on lib/index.js:
```diff
@@ -237,6 +245,11 @@ export default {
 						}
+					},
+					'appc:takeScreeshot': () => {
+						if (Project.isTitaniumApp || Project.isTitaniumModule) {
```
moved it below to `appc:open-or-close-related`

hansemannn commented on lib/menu/app.js:
```diff
@@ -14,6 +14,7 @@ module.exports = {
 						{ label: 'Toggle related files', command: 'appc:open-or-close-related' },
 						{ label: 'Add component...', command: 'appc:generate' },
 						{ label: 'Clean project', command: 'appc:clean' },
+						{ label: 'Screenshot of a device/simulator/emulator', command: 'appc:takeScreenshot' },
```
Simply "Take Screenshot"? More slicky for a menu command. Devs should know where they are trying to do the screenshot from, based on the run-config selected in the UI.

hansemannn commented on lib/index.js:
```diff
@@ -523,6 +532,36 @@ export default {
+		const filename = atom.config.get('appcelerator-titanium.general.screenshotPath') + "screenshot_" + new Date().getTime() + '.png';
+		const exec = require('child_process').exec;
+
+		if (os.platform() === 'darwin') {
```
Is this enough to determine if a Simulator is currently running? Some examples where this **could** fail:
- No Simulator running
- Build running for device, ad-hoc or production
- Running an Android / Windows build on a Mac

hansemannn replied:
@m1ga Should we help here? Let us know!

m1ga replied:
I think to check for all this would be some more work. E.g. on Android you could list the devices and check if there are multiple instances and use them in the screenshot command. On iOS I'm not sure how to check it. The current implementation would just not take a screenshot if the command doesn't exit with a success and display an error message.

Any help is of course very appreciated :smile:

hansemannn commented on lib/index.js:
```diff
@@ -523,6 +532,36 @@ export default {
+	 * Take a screenshot of a device/simulator/emulator
+	 */
+	takeScreenshot() {
+		const filename = atom.config.get('appcelerator-titanium.general.screenshotPath') + "screenshot_" + new Date().getTime() + '.png';
```
Mixed `"` and `'`. Also, what happens if no screenshot path is set?

hansemannn commented:
@m1ga I have added ae871ae to add some more validation around the plugin load state and iOS. Can you validate the changes? We also need to document the new option, doing that now.

m1ga commented:
Added `path.join()` to create the filename path cause I had a missing `/` at my machine. Can test the iOS part tomorrow. Android on Linux was working fine.

Also added a `-` in the command string so it will look better in the command palette

hansemannn commented:
Let me know how the tests run and it'll be ready to merge.

m1ga commented:
you can clone the repo, run `npm install` inside that folder and link it to  `.atom/packages/appc`
You could also use `apm` to install it `apm install` or `apm link` can't remember which one it was :) 

Tested it on OSX with an iOS simulator and no Android device connected -> added a small check to fix an error

hansemannn commented:
hehe, I know how to test it, I rather wanted to know your results. I think we are good to go then.

m1ga commented:
👍 

iOS with connected Android phone is working, too. 
But I had to restart Atom after attaching the phone while it still was running; "Refresh devices" was not able to find it before. But that might be a problem for a different ticket 😄
